Split contributing docs to multiple files (#36969)
mistercrunch/Airflow: contributing-docs/06_development_environments.rst

    with the License.  You may obtain a copy of the License at

 ..   http://www.apache.org/licenses/LICENSE-2.0

 .. Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

Development Environments
========================

There are two environments, available on Linux and macOS, that you can use to
develop Apache Airflow.

.. contents:: :local:

Local virtualenv Development Environment
----------------------------------------

All details about using and running local virtualenv environment for Airflow can be found
in `07_local_virtualenv.rst <07_local_virtualenv.rst>`__.

Benefits:

-   Packages are installed locally. No container environment is required.
-   You can benefit from local debugging within your IDE. You can follow the `Contributors quick start <contributor.rst#quick-start>`__
    to set up your local virtualenv and connect your IDE with the environment.

> -   With the virtualenv in your IDE, you can benefit from autocompletion and running tests directly from the IDE.

Reviewer: ```suggestion
-   With the virtualenv in your IDE, you can benefit from auto completion and running tests directly from the IDE.
```

    postgres database, hadoop, mongo, cassandra, redis, etc.

    The tests in Airflow are a mixture of unit and integration tests and some of
    them require these components to be set up. Local virtualenv supports only
    real unit tests. Technically, to run integration tests, you can configure
    and install the dependencies on your own, but it is usually complex.
    Instead, you are recommended to use
    `Breeze development environment <dev/breeze/doc/README.rst>`__ with all required packages
    pre-installed.

-   You need to make sure that your local environment is consistent with other
    developer environments. This often leads to a "works for me" syndrome. The
    Breeze container-based solution provides a reproducible environment that is
    consistent with other developers.

-   You are **STRONGLY** encouraged to also install and use `pre-commit hooks <08_static_code_checks.rst#pre-commit-hooks>`_
    for your local virtualenv development environment.
    Pre-commit hooks can speed up your development cycle a lot.

Typically you can connect your local virtualenv environments easily with your IDE
and use it for development:

> - `PyCharm/IntelliJ <quick-start-ide/contributors_quick_start_pycharm.rst>`__ quick start instrunctions

Reviewer: ```suggestion
- `PyCharm/IntelliJ <quick-start-ide/contributors_quick_start_pycharm.rst>`__ quick start instructions
```

- `VSCode <quick-start-ide/contributors_quick_start_vscode.rst>`__ quick start instructions

Breeze Development Environment
------------------------------

All details about using and running Airflow Breeze can be found in
`Breeze <dev/breeze/doc/README.rst>`__.

The Airflow Breeze solution is intended to ease your local development as "*It's
a Breeze to develop Airflow*".

Benefits:

-   Breeze is a complete environment that includes external components, such as
    mysql database, hadoop, mongo, cassandra, redis, etc., required by some of

>     Airflow tests. Breeze provides a preconfigured Docker Compose environment

Reviewer: ```suggestion
    Airflow tests. Breeze provides a pre-configured Docker Compose environment
```

| Reproducing CI failures  | - (-) not possible in many cases | - (+) fully reproducible              | (+) reproduce CI failures              |
+--------------------------+----------------------------------+---------------------------------------+----------------------------------------+
| Ability to update        | - (-) requires manual updates    | - (+) automated update via breeze cmd | (+/-) can be rebuild on demand         |
+--------------------------+----------------------------------+---------------------------------------+----------------------------------------+
| Disk space and CPU usage | - (+) relatively lightweight     | - (-) uses GBs of disk and many CPUs  | (-) integration tests (extra config)   |
+--------------------------+----------------------------------+---------------------------------------+----------------------------------------+
| IDE integration          | - (+) straightforward            | - (-) via remote debugging only       | (-) integration tests (extra config)   |
+--------------------------+----------------------------------+---------------------------------------+----------------------------------------+

> Typically, you are recommended to use both of these environments depending on your needs.

Reviewer: I believe the term "both" is for two options but the table described three setups?
```suggestion
Typically, you are recommended to use multiple of these environments depending on your needs.
```
